### PR TITLE
[CALCITE-1707] push down to druid all the extraction filters

### PR DESCRIPTION
**Pull request by b-slim**

### CODE REVIEW

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 1]
```diff
@@ -1421,7 +1421,6 @@ public Void apply(ResultSet resultSet) {
         + "    BindableFilter(condition=[AND(>=(/INT(Reinterpret($0), 86400000), 1997-01-01), <(/INT(Reinterpret($0), 86400000), 1998-01-01), OR(AND(>=(/INT(Reinterpret($0), 86400000), 1997-04-01), <(/INT(Reinterpret($0), 86400000), 1997-05-01)), AND(>=(/INT(Reinterpret($0), 86400000), 1997-06-01), <(/INT(Reinterpret($0), 86400000), 1997-07-01))))])\n"
         + "      DruidQuery(table=[[foodmart, foodmart]], intervals=[[1900-01-09T00:00:00.000/2992-01-10T00:00:00.000]], projects=[[$0]])";
     sql(sql)
-        .explainContains(explain)
```
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Why did you remove this?

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 2]
```diff
@@ -1695,28 +1694,120 @@ public Void apply(ResultSet resultSet) {
...
+        + "\"upper\":\"15\",\"upperStrict\":true,\"ordering\":\"numeric\"},"
+        + "{\"type\":\"selector\",\"dimension\":\"__time\",\"value\":\"1997\","
+        + "\"extractionFn\":{\"type\":\"timeFormat\",\"format\":\"yyyy\",\"timeZone\":\"UTC\","
+        + "\"locale\":\"en-US\"}}]},\"aggregations\":[{\"type\":\"doubleSum\","
+        + "\"name\":\"EXPR$0\",\"fieldName\":\"store_sales\"}],"
+        + "\"intervals\":[\"1900-01-09T00:00:00.000/2992-01-10T00:00:00.000\"],"
+        + "\"context\":{\"skipEmptyBuckets\":true}}";
     sql(sql)
         .queryContains(druidChecker(druidQuery))
-        .explainContains("PLAN=EnumerableInterpreter\n"
```
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Why? It would be good to still check the complete explain plan

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 3]
```diff
@@ -1695,28 +1694,120 @@ public Void apply(ResultSet resultSet) {
...
-        + "'intervals':['1900-01-09T00:00:00.000/2992-01-10T00:00:00.000'],"
-        + "'filter':{'type':'and',"
-        + "'fields':[{'type':'bound','dimension':'cases_per_pallet','lower':'8',"
-        + "'lowerStrict':false,'ordering':'numeric'},"
-        + "{'type':'bound','dimension':'cases_per_pallet','upper':'10','upperStrict':false,"
-        + "'ordering':'numeric'},{'type':'bound','dimension':'units_per_case','upper':'15',"
-        + "'upperStrict':true,'ordering':'numeric'}]},'dimensions':[],'metrics':['store_sales'],"
-        + "'granularity':'all','pagingSpec':{'threshold':16384,'fromNext':true},"
-        + "'context':{'druid.query.fetch':false}}";
+    String druidQuery = "{\"queryType\":\"timeseries\",\"dataSource\":\"foodmart\","
```
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Is it possible to use single quotes for DruidQuery instead of escaping the double quotes?

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java [hunk 4]
```diff
@@ -1695,28 +1694,120 @@ public Void apply(ResultSet resultSet) {
...
         .returnsUnordered("EXPR$0=75364.09998679161");
   }
+
+  @Test public void testPushOfFilterExtractionOnDayAndMonth() {
+    String sql = "SELECT \"product_id\" , EXTRACT(day from \"timestamp\"), EXTRACT(month from "
+        + "\"timestamp\") from \"foodmart\" WHERE  EXTRACT(day from \"timestamp\") >= 30 AND "
+        + "EXTRACT(month from \"timestamp\") = 11 "
+        + "AND  \"product_id\" >= 1549 group by \"product_id\", EXTRACT(day from "
+        + "\"timestamp\"), EXTRACT(month from \"timestamp\")";
+    sql(sql)
```
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: We should add the DruidQuery in this case too, even if we verify that we are actually pushing the EXTRACT for other queries below.

File: druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Utility class for extraction function mapping between SQL and Druid
+ */
+public class ExtractionFunctionUtil {
+
+  private ExtractionFunctionUtil() {
+  }
+
+  /**
+   * @param rexNode RexNode that might contains an extraction function
```
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java]: A short description would be nice.

File: druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+
+
+/**
+ * Utility class for extraction function mapping between SQL and Druid
+ */
+public class ExtractionFunctionUtil {
```
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java]: You can make the class _final_.

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java [hunk 7]
```diff
@@ -205,12 +205,12 @@ public void onMatch(RelOptRuleCall call) {
         intervals = DruidDateTimeUtils.createInterval(
             query.getRowType().getFieldList().get(timestampFieldIdx).getType(),
             RexUtil.composeConjunction(rexBuilder, triple.getLeft(), false));
-        if (intervals == null) {
-          // We can't push anything useful to Druid.
-          residualPreds.addAll(triple.getLeft());
+        if (intervals == null || intervals.isEmpty()) {
+          // Case we have an filter with extract that can not be written as interval push down
+          triple.getMiddle().addAll(triple.getLeft());
```
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java]: Cool!

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java [hunk 8]
```diff
@@ -980,35 +997,43 @@ private JsonFilter translateFilter(RexNode e) {
         final boolean numeric =
             call.getOperands().get(posRef).getType().getFamily()
                 == SqlTypeFamily.NUMERIC;
+
+        final ExtractionFunction extractionFunction = ExtractionFunctionUtil.buildExtraction(call
+            .getOperands().get(posRef));
+        String dimName = tr(e, posRef);
+        if (dimName.equals(DruidConnectionImpl.DEFAULT_RESPONSE_TIMESTAMP_COLUMN)) {
+          dimName = "__time";
```
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java]: Please, use DruidTable.DEFAULT_TIMESTAMP_COLUMN here and leave a comment explaining why we are renaming the dimension column.

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java [hunk 9]
```diff
@@ -205,12 +205,12 @@ public void onMatch(RelOptRuleCall call) {
...
         }
       }
-      if (intervals == null && triple.getMiddle().isEmpty()) {
+      if ((intervals == null || intervals.isEmpty()) && triple.getMiddle().isEmpty()) {
```
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java]: This _if_ block can be removed now: we will never enter.

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 3 above)
Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: does it make any difference ?

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java (hunk 9 above)
Comment (b-slim) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java]: i am sure we have cases where the intervals are empty so not sure what you mean by removing it ?

File: druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java (hunk 6 above)
Comment (b-slim) [inline druid/src/main/java/org/apache/calcite/adapter/druid/ExtractionFunctionUtil.java]: ok

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java (hunk 8 above)
Comment (b-slim) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java]: ok

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 4 above)
Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: ok

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 2 above)
Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: the druid query has all the plan, so not sure why have both ? i think one is enough if the query is fully executed by druid.

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 1 above)
Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: good catch

File: druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java (hunk 9 above)
Comment (jcamachor) [inline druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java]: This condition will never be true: if intervals are null or empty, we will have put the corresponding predicates in triple.getMiddle. If they are empty because triple.getLeft was empty and triple.getMiddle is empty too, we will have entered in L198.

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 3 above)
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Readability + having uniform code base.

File: druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java (hunk 2 above)
Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: How do we know we do not have any computation in Calcite side if we do not check the explain plan? For instance, if there is a bug in the rewriting rules, we might push computation to Druid and still leave (useless) operators on top of the DruidQuery?

Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: because i read the druid query and it matches exactly the sql query this means all is pushed and the query make sense, so not sure what is the issue here ?

Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: We have a rule that can push an operator _partially_ into Druid, i.e., some computation is pushed to Druid and some part stays in the plan. I want to double check that we do not compute anything twice, as _pushing everything to Druid_ is not equal to _not leaving the same operation in the plan_. Why don't we add the explain plan check to extend the test coverage and be certain now and in the future that this is working correctly, even if I get your point that current rule should not be doing that?

Comment (b-slim) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: i will add it to please you.

Comment (jcamachor) [inline druid/src/test/java/org/apache/calcite/test/DruidAdapterIT.java]: Thanks

**Comment by b-slim** [comment]
@jcamachor  thanks for the review, i think i have  fixed all of it.
